src/modelfree/common/utils.py

```diff
@@ -149,39 +155,114 @@ def _reset_video_recorder(self, metadata=None):
+        self.num_envs = num_envs
+        self.num_policies = num_policies
+        self.num_traj_to_save = num_traj_to_save
+        self.traj_dir = traj_dir
+        if not os.path.isdir(self.traj_dir):
+            os.makedirs(self.traj_dir)
+
+        self.traj_dicts = [[defaultdict(list) for e in range(num_envs)]
+                           for p in range(num_policies)]
+        self.full_traj_dicts = [defaultdict(list) for p in range(num_policies)]
+        self.num_completed = 0
+        self.already_saved = [False for p in range(num_policies)]
+
+    def record_traj(self, rewards, actions, dones, prev_observations):
+        data_keys = ('rewards', 'actions', 'obs')
+        data_vals = (rewards, actions, prev_observations)
+        iter_space = itertools.product(enumerate(self.traj_dicts), range(self.num_envs))
+        # iterate over both agents over all environments in VecEnv
+        for (agent_idx, agent_dicts), env_idx in iter_space:
+            for key, val in zip(data_keys, data_vals):
+                agent_dicts[env_idx][key].append(val[agent_idx][env_idx])
+            if dones[env_idx]:
+                ep_len = len(agent_dicts[env_idx]['rewards'])
+                ep_starts = [True] + [False] * (ep_len - 1)
+                self.full_traj_dicts[agent_idx]['episode_starts'].append(np.array(ep_starts))
+
+                ep_ret = sum(agent_dicts[env_idx]['rewards'])
+                self.full_traj_dicts[agent_idx]['episode_returns'].append(np.array([ep_ret]))
+
+                for key in data_keys:
+                    # consolidate episode data and append to long-term data dict
+                    episode_key_data = np.array(agent_dicts[env_idx][key])
+                    self.full_traj_dicts[agent_idx][key].append(episode_key_data)
+                agent_dicts[env_idx] = defaultdict(list)
+                # so as not to double-count number of episodes
+                self.num_completed += int(agent_idx == 0)
+                if self.num_completed >= self.num_traj_to_save:
+                    self.save_traj(agent_idx)
+
+    def save_traj(self, agent_idx):
+        if self.already_saved[agent_idx]:
+            print(f'Warning: already saved trajectories for agent {agent_idx}')
+        dump_dict = {
+            k: np.concatenate(v, axis=0)
+            for k, v in self.full_traj_dicts[agent_idx].items()}
+        save_path = os.path.join(self.traj_dir, f'agent_{agent_idx}.npz')
+        np.savez(save_path, **dump_dict)
+        self.already_saved[agent_idx] = True
+
+
+def simulate(venv, policies, render=False, record_traj=False, traj_dir='data/',
+             num_traj_to_save=None):
     """
     Run Environment env with the agents in agents
     :param venv(VecEnv): vector environment.
     :param policies(list<BaseModel>): a policy per agent.
     :param render: true if the run should be rendered to the screen
+    :param record_traj: whether to save trajectory data in stable_baselines.GAIL format
```

Comment:
If we make the `TrajectoryRecorder` a wrapper, then we can move this out of `simulate` into the constructor params.